### ISSUE TITLE
[UPDATE] Fix failing tests after new NF DSL2 syntax update

### ISSUE THREAD

**Issue by drpatelh** [nf-core/modules#1094]
Please submit PRs fixing these modules to my fork:
https://github.com/drpatelh/nf-core-modules/tree/update

You should be able to run `pytest` as normal but the options are now in a `nextflow.config` like [here](https://github.com/drpatelh/nf-core-modules/blob/update/tests/modules/star/align/nextflow.config) instead of using the `addParams`

## Failing at module level

**Still failing**
- [ ] unicycler

## Fixed at module level

- [x] arriba (@JoseEspinosa)
- [x] bcftools/query (@JoseEspinosa )
- [x] bowtie/align (@drpatelh)
- [x] bowtie2/align (@drpatelh)
- [x] bwa/aln (@drpatelh)
- [x] bwa/mem (@drpatelh)
- [x] bwamem2/mem (@drpatelh)
- [x] clonalframeml (@drpatelh) (@rpetit3)
- [x] cooler/merge (@JoseEspinosa)
- [x] cooler/zoomify (@JoseEspinosa)
- [x] csvtk/split (@JoseEspinosa)
- [x] delly/call (@JoseEspinosa)
- [x] ensemblvep (tests pass on Docker and Singularity)
- [x] fgbio/fastqtobam (@drpatelh)
- [x] freebayes (@maxulysse)
- [x] gatk4/intervallisttools (@JoseEspinosa)
- [x] gatk4/variantfiltration (@JoseEspinosa)
- [x] glnexus (@drpatelh)
- [x] gstama/collapse (@JoseEspinosa)
- [x] hisat2/align (@drpatelh)
- [x] hicap  (@drpatelh) (@rpetit3)
- [x] hisat2/build (@drpatelh)
- [x] imputeme/vcftoprs (@lassefolkersen ) (Singularity timeout)
- [x] jupyternotebook (@grst)
- [x] kallistobustools/count (@JoseEspinosa)
- [x] kallistobustools/ref (@JoseEspinosa)
- [x] leehom (@JoseEspinosa)
- [x] maltextract
- [x] metaphlan3 (@JoseEspinosa)
- [x] ncbigenomedownload
- [x] pairtools/select (@JoseEspinosa)
- [x] pirate  (@drpatelh) (@rpetit3) (fixed test data but need to update `test.yml`)(#1444)
- [x] plink2/vcf (@drpatelh)
- [x] pydamage/filter (@JoseEspinosa)
- [x] rmarkdownnotebook (@grst)
- [x] roary (@rpetit3) (#1418)
- [x] rsem/calculateexpression (@drpatelh)
- [x] salmon/quant (@drpatelh)
- [x] snpeff (#1877)
- [x] sratools/fasterqdump (@drpatelh)
- [x] tbprofiler/profile  (@rpetit3 )
- [x] unicycler (@JoseEspinosa)
- [x] yara/mapper (@drpatelh)

## Already passing on local tests

- [x] chromap/chromap (@drpatelh)
- [x] gatk4/mergebamalignment (@JoseEspinosa)
- [x] fgbio/groupreadsbyumi (@drpatelh)
- [x] nanoplot (@JoseEspinosa)
- [x] plasmidid (@drpatelh)
- [x] prokka (@JoseEspinosa) (worked with Docker but conda env was not resolved)
- [x] shovill (@JoseEspinosa) (These two relative big [test files](https://github.com/nf-core/modules/blob/fcbf5a95834f11fc9e5b70b1e38367a9bec29df5/tests/modules/shovill/main.nf#L12-L13) do not stage properly sometimes and tests break then)

## Failing at sub-workflow level

- [x] samtools/flagstat (@drpatelh)
- [x] samtools/idxstats (@drpatelh)
- [x] samtools/index (@drpatelh)
- [x] samtools/sort (@drpatelh)
- [x] samtools/stats (@drpatelh)
- [x] sratools/prefetch (@drpatelh)

**Comment by JoseEspinosa** [comment]
`kallistobustools/ref` was giving this error (actually also `kallistobustools/count`):

```
Launching `./tests/modules/kallistobustools/ref/main.nf` [sad_marconi] - revision: 59ac0a19e6
[-        ] process > test_kallistobustools_ref:KALLISTOBUSTOOLS_REF -
No such variable: containerEngine
```
I fixed this by changing the container directive see [here](https://github.com/JoseEspinosa/nf-core-modules/commit/3819cbb20fa9094ef3926b66a9dda623cca2ae32) 
Actually, I populate the module template in the same way see [here](https://github.com/nf-core/tools/blob/dbeee2b39b91cc1a1d4fd5d32c6131638ec02997/nf_core/module-template/modules/main.nf#L27-L30) and I was wondering if we should update all the modules again with this... 😢  ping @drpatelh

**Comment by drpatelh** [comment]
Hmmm....why is it not working for that module in particular when it is for all of the others? Think we need to figure that out.

**Comment by JoseEspinosa** [comment]
The difference can be seen in the commit I provided in the comment above:
The original code was:

```
container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
     'quay.io/biocontainers/kb-python:0.26.3--pyhdfd78af_0' }"
```
The code that solves the problem is:

```
container workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
             'quay.io/biocontainers/kb-python:0.26.3--pyhdfd78af_0'
```
The only difference I can spot with other modules is the dash in `kb-python` but shouldn't make any difference since is within single quotes... 🤔

**Comment by drpatelh** [comment]
This is really weird. @mahesh-panchal any idea why having the closure would cause this problem?

**Comment by JoseEspinosa** [comment]
The reason is that in the workflow definition in the test `main.nf` there is a variable defined as `workflow = standard`: https://github.com/drpatelh/nf-core-modules/blob/fcbf5a95834f11fc9e5b70b1e38367a9bec29df5/tests/modules/kallistobustools/ref/main.nf#L11
Although I don't understand very well why this is not happening when I modified the container directive 🤔

**Comment by JoseEspinosa** [comment]
The reason is that there is an input value channel defined as `workflow` in the module  `main.nf`: https://github.com/drpatelh/nf-core-modules/blob/3819cbb20fa9094ef3926b66a9dda623cca2ae32/modules/kallistobustools/ref/main.nf#L13
However, I don't understand why changing the container directive was solving the problem... 🤔

**Comment by drpatelh** [comment]
Ahhh. So it's a scope issue  The closure evaluates any variables in the local scope as priority? So we can either change the input value to something else or remove the closure if it isn't required?

**Comment by JoseEspinosa** [comment]
I updated the input value to `workflow_mode`. Regarding defining an input value called `task` I made the test and seems to work but gives a warning:

```
[a9/b8a539] process > test_kallistobustools_count:KALLISTOBUSTOOLS_COUNT (test) [100%] 1 of 1 ✔
WARN: Process test_kallistobustools_count:KALLISTOBUSTOOLS_COUNT overrides reserved variable `task`
```

**Comment by drpatelh** [comment]
I would still be in favour of removing the closure if we don't need it for simplicity. Let's see what @mahesh-panchal thinks.

**Comment by lassefolkersen** [comment]
I checked the`imputeme/vcftoprs` module on an AWS instance, using latest nextflow, pullling like this 
``` 
```
PROFILE=docker nextflow run tests/modules/imputeme/vcftoprs -entry test_imputeme_vcftoprs -c tests/config/nextflow.config
```
and this
```
NF_CORE_MODULES_TEST=1 PROFILE=docker pytest --tag imputeme/vcftoprs --symlink --wt 2 --keep-workflow-wd
```

And there's no failures. Run time 6m 38s. Is it too slow?

**Comment by grst** [comment]
It took 21:15 on my system, but I think most of the time was used to pull
the singularity container.

**Comment by drpatelh** [comment]
Yes, I saw the same thing:

```
Caused by:
  Failed to pull singularity image
  command: singularity pull  --name containers.biocontainers.pro-s3-SingImgsRepo-imputeme-vv1.0.7_cv1-imputeme_vv1.0.7_cv1.img.img.pulling.1637839133887 https://containers.biocontainers.pro/s3/SingImgsRepo/imputeme/vv1.0.7_cv1/imputeme_vv1.0.7_cv1.img > /dev/null
  status : 143
  message:
    INFO:    Downloading network image
```

Which is weird because you can manually download by clicking the link.

**Comment by grst** [comment]
anyway, `vcftoprs` seems to work, it was just a timeout issue during the bulk tests.

**Comment by mahesh-panchal** [comment]
> I would still be in favour of removing the closure if we don't need it for simplicity. Let's see what @mahesh-panchal thinks.

There's no closure there. It's only an expression evaluation. If it works outside the string then leave it out. I only tested the string version because `task.ext` couldn't be seen in an if statement, but I knew it worked it that context.

**Comment by drpatelh** [comment]
We couldn't get it working consistently without the evaluation anyway so will leave things as they are and update later if required.

**Comment by rpetit3** [comment]
@JoseEspinosa you are correct  on `ncbigenomedownload` he test is downloading everything. The `addParams` was important for the test, we'll have add them in an alternate way or modify the recipe a little.

Originally the test looked like this
```
include { NCBIGENOMEDOWNLOAD } from '../../../modules/ncbigenomedownload/main.nf' addParams( options: [ args: '-A GCF_000013425.1 --formats genbank,fasta,assembly-stats bacteria '] )
```

**Comment by JoseEspinosa** [comment]
@rpetit3 they are now passed with [this](https://github.com/drpatelh/nf-core-modules/blob/c18bb015efde4440f12e57b9a553b4f607360ff9/tests/modules/ncbigenomedownload/nextflow.config) `nextflow.config`. I just checked the `.command.sh` and they are actually being passed:

```
$ cat .command.sh
#!/bin/bash -ue
ncbi-genome-download \
    -A GCF_000013425.1 --formats genbank,fasta,assembly-stats bacteria \
     \
    --output-folder ./ \
    --flat-output

cat <<-END_VERSIONS > versions.yml
test_ncbigenomedownload:NCBIGENOMEDOWNLOAD:
    ncbigenomedownload: $( ncbi-genome-download --version )
END_VERSIONS
```
Any ideas?

**Comment by rpetit3** [comment]
You're right! Apologies completely missed that.

**Comment by rpetit3** [comment]
Happy to report PIRATE is good again!

**Comment by muffato** [comment]
All but unicycler actually got fixed at some point since this issue was open.
Since unicycler is already reported as needing a fix in #2154 , I'll close this issue here. No need to track it in two places
